apiserver/storageprovisioner/internal/filesystemwatcher/watchers.go

```diff
@@ -0,0 +1,453 @@
+		changes:                make(set.Strings),
+		machineFilesystems:     fw.Backend.WatchMachineFilesystems(m),
+		modelFilesystems:       fw.Backend.WatchModelFilesystems(),
+		modelVolumeAttachments: fw.Backend.WatchModelVolumeAttachments(),
+		modelVolumesAttached:   make(set.Tags),
+		modelVolumeFilesystems: make(map[names.VolumeTag]names.FilesystemTag),
+	}
+	go func() {
+		defer w.tomb.Done()
+		defer watcher.Stop(w.machineFilesystems, &w.tomb)
+		defer watcher.Stop(w.modelFilesystems, &w.tomb)
+		defer watcher.Stop(w.modelVolumeAttachments, &w.tomb)
+		w.tomb.Kill(w.loop())
+	}()
+	return w
+}
+
+// machineFilesystemsWatcher is a strings watcher that reports both
+// machine-scoped filesystems, and model-scoped, volume-backed filesystems
+// that are attached to the specified machine.
+//
+// NOTE(axw) we use the existence of the *volume* attachment rather than
+// filesystem attachment because the filesystem attachment can be destroyed
+// before the filesystem, but the volume attachment cannot.
+type machineFilesystemsWatcher struct {
+	stringsWatcherBase
+	changes                set.Strings
+	backend                Backend
+	machine                names.MachineTag
+	machineFilesystems     state.StringsWatcher
+	modelFilesystems       state.StringsWatcher
+	modelVolumeAttachments state.StringsWatcher
+	modelVolumesAttached   set.Tags
+	modelVolumeFilesystems map[names.VolumeTag]names.FilesystemTag
+}
+
+func (w *machineFilesystemsWatcher) loop() error {
+	defer close(w.out)
+	var out chan<- []string
+	var machineFilesystemsReceived bool
+	var modelFilesystemsReceived bool
+	var modelVolumeAttachmentsReceived bool
+	var sentFirst bool
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		case values, ok := <-w.machineFilesystems.Changes():
+			if !ok {
+				return watcher.EnsureErr(w.machineFilesystems)
+			}
+			machineFilesystemsReceived = true
+			for _, v := range values {
+				w.changes.Add(v)
+			}
+		case values, ok := <-w.modelFilesystems.Changes():
+			if !ok {
+				return watcher.EnsureErr(w.modelFilesystems)
+			}
+			modelFilesystemsReceived = true
+			for _, id := range values {
+				filesystemTag := names.NewFilesystemTag(id)
+				if err := w.modelFilesystemChanged(filesystemTag); err != nil {
+					return errors.Trace(err)
+				}
+			}
+		case values, ok := <-w.modelVolumeAttachments.Changes():
+			if !ok {
+				return watcher.EnsureErr(w.modelVolumeAttachments)
+			}
+			modelVolumeAttachmentsReceived = true
+			for _, id := range values {
+				machineTag, volumeTag, err := state.ParseVolumeAttachmentId(id)
+				if err != nil {
+					return errors.Annotate(err, "parsing volume attachment ID")
+				}
+				if machineTag != w.machine {
+					continue
+				}
+				if err := w.modelVolumeAttachmentChanged(volumeTag); err != nil {
+					return errors.Trace(err)
+				}
+			}
+		case out <- w.changes.SortedValues():
+			w.changes = make(set.Strings)
+			out = nil
+		}
+		if machineFilesystemsReceived &&
```

Comment:
Yes. The way all watchers work is to first give the initial state of the world in a single event, and then later changes can report on individual aspects. If we were to report the machine-scoped filesystems without also reporting the model-scoped ones, then the worker would think that the model-scoped ones were not part of the model any more.